pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,290 @@
+
+                        if kf == 'field' and kf not in self._expr:
+                            raise KeyError(f"Missing field '{kf}' used in pipeline")
+                pl[f'pl{i}/f{j}/params'] = params
+
+        return pipel_cfg
+
+    def _eval_exprs(self, intg):
+        # Get the primitive variable names
+        pnames = self.elementscls.privarmap[self.ndims]
+
+        # Compute the gradients
+        if self._gradpinfo:
+            grad_soln = intg.grad_soln
+
+        # Iterate over each element type in the simulation
+        for i, (soln, (ename, eles)) in enumerate(zip(intg.soln, 
+                                                  self._ele_map.items())):
+            
+            dom_str = f'domain_{self.dom_0 + i}/'
+            # Subset and transpose the solution
+            soln = soln[...,:].swapaxes(0, 1)
+
+            mesh_op, soln_op = self.ops[ename]
+            # Interpolate the solution to the subdivided points
+            soln = soln_op @ soln
+
+            # Convert from conservative to primitive variables
+            psolns = self.elementscls.con_to_pri(soln, self.cfg)
+
+            # Prepare the substitutions dictionary
+            subs = dict(zip(pnames, psolns), t=intg.tcurr)
+            # subs |= dict(zip('xyz', plocs)) # TODO: add the ability for xyz in functions
+
+            # Prepare any required gradients
+            if self._gradpinfo:
+                grads = np.rollaxis(grad_soln[idx], 2)[..., :]
+
+                # Interpolate the gradients to the points
+                grad_soln = soln_op @ grad_soln
+
+                # Transform from conservative to primitive gradients
+                pgrads = self.elementscls.grad_con_to_pri(soln, grads,
+                                                          self.cfg)
+
+                # Add them to the substitutions dictionary
+                for pname, idx in self._gradpinfo:
+                    for dim, grad in zip('xyz', pgrads[idx]):
+                        subs[f'grad_{pname}_{dim}'] = grad
+
+            for field in self._expr:
+                # Evaluate the expression at each point
+                fun = npeval(self._expr[field], subs)
+                self.mesh_n[dom_str + f'fields/{field}/values'] = fun.T
+
+    def _get_expr(self, expr, subs={}):
+        # Ensure the expression does not contain invalid characters
+        if not re.match(r'[A-Za-z0-9_ \t\n\r.,+\-*/%()]+$', expr):
+            raise ValueError('Invalid characters in expression')
+
+        # Substitute variables
+        if subs:
+            expr = re.sub(r'\b({0})\b'.format('|'.join(subs)),
+                          lambda m: str(subs[m[1]]), expr)
+
+        # Convert integers not inside [] to floats
+        expr = re.sub(r'\b((\d+\.?\d*)|(\.\d+))([eE][+-]?\d+)?(?![^[]*\])',
+                      _ensure_float, expr)
+
+        # Encase in parenthesis
+        return f'({expr})'
+
+    def _get_mesh_op(self, name, nspts, svpts):
```

Comment:
Should be able to inline all of these like what the integration type plugins do.